File: app/aircraft_db.py
"""Aircraft registry / type lookup (ICAO24 → registration + type).

Reads a gzipped semicolon-separated CSV in the tar1090-db / Mictronics shape:

    icao24_hex;registration;type_icao;flags;type_long;;;

Looked up by the hex ICAO24 address we already track. Loading is opt-in —
if no DB file is present the registry stays empty and enrichment silently
no-ops.
"""

import csv
import gzip
import logging
from pathlib import Path

log = logging.getLogger("beast.aircraft_db")

# Checked in order; first existing file wins. The /data path lets users drop
# a fresher DB in via the volume mount; the in-image path is the fallback
# baked at build time (if the Dockerfile downloads one).
DEFAULT_PATHS: list[Path] = [
    Path("/data/aircraft_db.csv.gz"),
    Path(__file__).parent / "aircraft_db.csv.gz",
]


class AircraftDB:
    """In-memory ICAO24 → registration/type lookup."""

    def __init__(self) -> None:
        self._db: dict[str, dict[str, str | None]] = {}

    def __len__(self) -> int:
        return len(self._db)

    def __contains__(self, icao: str) -> bool:
        return icao.lower() in self._db

    def lookup(self, icao: str) -> dict[str, str | None] | None:
        return self._db.get(icao.lower())

    def load_from(self, path: Path) -> int:
        """Load entries from a gzipped semicolon-separated CSV. Returns row count."""
        with gzip.open(path, "rt", encoding="utf-8", errors="replace") as fh:
            reader = csv.reader(fh, delimiter=";")
            for row in reader:
                if len(row) < 5 or not row[0]:
                    continue
                icao = row[0].strip().lower()
                reg = (row[1] or "").strip() or None
                type_icao = (row[2] or "").strip() or None
                type_long = (row[4] or "").strip() or None
                if reg or type_icao or type_long:
                    self._db[icao] = {
                        "registration": reg,
                        "type_icao": type_icao,
                        "type_long": type_long,
                    }
        return len(self._db)

    def load_first_available(self, paths: list[Path] | None = None) -> int:
        """Try each candidate path in order; load the first one that exists."""
        for p in paths or DEFAULT_PATHS:
            if p.exists():
                try:
                    n = self.load_from(p)
                    log.info("loaded aircraft DB from %s (%d entries)", p, n)
                    return n
                except Exception as e:
                    log.warning("failed to load aircraft DB from %s: %s", p, e)
        log.info("no aircraft DB found; enrichment disabled")
        return 0

File: dotnet/src/FlightJar.Core/ReferenceData/AircraftDb.cs
using System.Collections.Frozen;
using System.IO.Compression;
using FlightJar.Core.State;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace FlightJar.Core.ReferenceData;

/// <summary>
/// ICAO24 → registration/type lookup backed by a gzipped semicolon-separated
/// CSV in the <a href="https://github.com/wiedehopf/tar1090-db">tar1090-db /
/// Mictronics</a> shape. Mirrors <c>app/aircraft_db.py</c>. Load is opt-in —
/// if no file is present, lookups no-op and snapshot enrichment stays empty.
/// </summary>
public sealed class AircraftDb : IAircraftDb
{
    private readonly ILogger _logger;
    private FrozenDictionary<string, AircraftDbEntry> _entries = FrozenDictionary<string, AircraftDbEntry>.Empty;

    public AircraftDb(ILogger<AircraftDb>? logger = null)
    {
        _logger = logger ?? NullLogger<AircraftDb>.Instance;
    }

    public int Count => _entries.Count;

    public AircraftDbEntry? Lookup(string icao)
    {
        if (string.IsNullOrEmpty(icao))
        {
            return null;
        }
        return _entries.TryGetValue(icao.ToLowerInvariant(), out var e) ? e : null;
    }

    /// <summary>Parse a gzipped CSV and swap it in atomically on success.</summary>
    public async Task<int> LoadFromAsync(string path, CancellationToken ct = default)
    {
        var fresh = new Dictionary<string, AircraftDbEntry>(StringComparer.Ordinal);
        await using var file = File.OpenRead(path);
        await using var gz = new GZipStream(file, CompressionMode.Decompress);
        using var reader = new StreamReader(gz);
        string? line;
        while ((line = await reader.ReadLineAsync(ct)) is not null)
        {
            if (line.Length == 0)
            {
                continue;
            }
            var parts = line.Split(';');
            if (parts.Length < 5 || parts[0].Length == 0)
            {
                continue;
            }
            var icao = parts[0].Trim().ToLowerInvariant();
            if (icao.Length == 0)
            {
                continue;
            }
            var reg = NullIfEmpty(parts[1]);
            var typeIcao = NullIfEmpty(parts[2]);
            var typeLong = NullIfEmpty(parts[4]);
            if (reg is null && typeIcao is null && typeLong is null)
            {
                continue;
            }
            fresh[icao] = new AircraftDbEntry(reg, typeIcao, typeLong);
        }
        _entries = fresh.ToFrozenDictionary(StringComparer.Ordinal);
        return _entries.Count;
    }

    /// <summary>Try each candidate path; load the first one that exists.</summary>
    public async Task<int> LoadFirstAvailableAsync(IEnumerable<string> paths, CancellationToken ct = default)
    {
        foreach (var p in paths)
        {
            if (!File.Exists(p))
            {
                continue;
            }
            try
            {
                var n = await LoadFromAsync(p, ct);
                _logger.LogInformation("loaded aircraft DB from {Path} ({Count} entries)", p, n);
                return n;
            }
            catch (Exception ex) when (ex is not OperationCanceledException)
            {
                _logger.LogWarning(ex, "failed to load aircraft DB from {Path}", p);
            }
        }
        _logger.LogInformation("no aircraft DB found; enrichment disabled");
        return 0;
    }

    private static string? NullIfEmpty(string s)
    {
        var trimmed = s.Trim();
        return trimmed.Length == 0 ? null : trimmed;
    }
}
